lua/neotest-moonbit/init.lua

```diff
@@ -0,0 +1,254 @@
+local lib = require("neotest.lib")
+local types = require("neotest.types")
+local logger = require("neotest.logging")
+
+local jsonlist = require("neotest-moonbit.json")
+
+local joinpath = vim.fs.joinpath
+local basename = vim.fs.basename
+
+local filetype = require("plenary.filetype")
+if filetype.detect_from_extension("x.mbt") == "" then
+  filetype.add_table({ extension = { mbt = "moonbit" } })
+end
+
+--- @class Adapter : neotest.Adapter
+local M = {}
+
+local MOON_MOD_JSON = "moon.mod.json"
+local MOON_PKG_JSON = "moon.pkg.json"
+
+M.Adapter = { name = "neotest-moonbit" }
+
+---Find the project root directory given a current directory to work from.
+---Should no root be found, the adapter can still be used in a non-project context if a test file matches.
+---@async
+---@param dir string @Directory to treat as cwd
+---@return string | nil @Absolute root dir of test suite
+function M.Adapter.root(dir)
+  -- we want to ensure we are sitting at project root
+  if dir == lib.files.match_root_pattern(MOON_MOD_JSON)(dir) then
+    return dir
+  end
+  return nil
+end
+
+local function readJSON(path)
+  local f, err = io.open(path, "r"):read("*a")
+  if err then
+    logger.error(err)
+    return nil
+  end
+  local ok, json = pcall(vim.json.decode, f)
```

Comment:
```suggestion
  local f, err = io.open(path, "r")
  if err then
    logger.error(err)
    return nil
  end
  local content = f:read("*a")
  local ok, json = pcall(vim.json.decode, content)
```

As for my testing, using `io.open(..):read()` cause the readJSON to throw an error